### DISCUSSION TITLE
Rebalancing extremely slow

### DISCUSSION THREAD

**Discussion by mandrew1983** [moosefs/moosefs#447] [Q&A]
Hello,

I just want to get confirm if this behavior is normal and if not, I'm hoping I can be pointed at the right direction on what to check. I installed the bottom 3 chunk servers on this image and after about 3 weeks, this is its status:

![image](https://user-images.githubusercontent.com/87539561/131165477-a4636bd0-7888-4687-bf56-cec340123a6f.png)

I left the replication speeds config to default. The config is similar to that of the first 2 and I do not see anything unusual on the server logs.

Thank you in advance for the assistance.

**Comment by chogata** [comment]
Replication speed is governed by replication limits and if it is constant-ish in time, then it works okay. Can you post screenshot of your master charts in short and medium range (the deletions and replications charts are the 2 important here, so make sure to include them).

**Reply by mandrew1983** [reply]
Thank you for your reply. Here are the screenshots:

![image](https://user-images.githubusercontent.com/87539561/131511826-81d2a194-e0d9-472a-9c9a-3838048daebb.png)
![image](https://user-images.githubusercontent.com/87539561/131511967-7845eeb3-006e-4bf3-8761-37331c9503ff.png)

Let me know if you need additional information. Also please bear with me as I am still learning to manage MFS.

**Reply by chogata** [reply]
Hi,
This is short range screenshot only, it's not a lot of information in itself. Medium and even long range is needed for complete picture of the situation. Click on the "medium range" and "long range" link above the charts and make screenshots in both settings.

**Reply by mandrew1983** [reply]
Ah apologies.

Medium range:
![image](https://user-images.githubusercontent.com/87539561/132550574-1f32b809-ddcc-43e4-8b26-bad91e6b5d35.png)

Long range:
![image](https://user-images.githubusercontent.com/87539561/132550698-4e368995-b4c9-4807-925c-165f87c2e558.png)

Very long range:
![image](https://user-images.githubusercontent.com/87539561/132550770-92a7179b-8cbc-483e-9e0d-41d73910d19d.png)

Thank you again and apologies for the very late reply. I had no internet access for the past week.

**Comment by chogata** [comment]
OK, so from the medium and long range, it looks like the replication speed did not decrease over time, it was just like this. Your chunk servers are quite big, 175 TiB space on one is in the upper range of what we see. Now, a very important question: how are the physical hard drives configured? Is it JBOD, or do you have a logical array? If this is the latter, you did not read the instruction ;) If this is the former, you can play around with replications limits. Basically, those 2 config values are of interest here:

```
# Maximum number of chunks to replicate to one chunkserver (default is 2,1,1,4)
# one number is equal to four same numbers separated by colons
#  first limit is for endangered chunks (chunks with only one copy)
#  second limit is for undergoal chunks (chunks with number of copies lower than specified goal)
#  third limit is for rebalance between servers with space usage around arithmetic mean
#  fourth limit is for rebalance between other servers (very low or very high space usage)
# usually first number should be grater than or equal to second, second greater than or equal to third, and fourth greater than or equal to third ( 1st >= 2nd >= 3rd <= 4th )
# CHUNKS_WRITE_REP_LIMIT = 2,1,1,4

# Maximum number of chunks to replicate from one chunkserver (default is 10,5,2,5)
# one number is equal to four same numbers separated by colons
# limit groups are the same as in write limit, also relations between numbers should be the same as in write limits ( 1st >= 2nd >= 3rd <= 4th )
# CHUNKS_READ_REP_LIMIT = 10,5,2,5
```

Uncomment the `CHUNKS_WRITE_REP_LIMIT` and `CHUNKS_READ_REP_LIMIT` and adjust the 4th value. You will have to experiment. For sure neither of them should be greater than the number of disks in your chunk server machine, BUT if you have like 30 disks, setting it to 30 might not be a good idea, because the rest of your hardware might not be able to keep up. I would start with doubling it and observing the system. If your processor usage on chunk server machines doesn't increase drastically, you are not observing any timeouts in your installation and the clients work okay, you can try to add another fraction to the values (do not double the second time, just add the original values again or even less). If at any point you observe any problems, go back to the previous value.

If you do not have JBOD but a logical array, you can of course also increase the replication speed, but it's hard to advise how much and what to look out for in your array controller, to see if it is not overloaded with i/o operations and if single drives are not overly used.

Whatever you do, just keep in mind this number indicates how many simultaneous operations of certain type (read or write) the master is allowed to send to any single chunk server.

**Reply by mandrew1983** [reply]
I do have a JBOD architecture for my chunk servers. Each chunks server in question have a total of 12 physical hard drives on each one, not counting the 2 hard drives for the OS. 

I will make the changes you suggested. And just to confirm, you are talking about the values that are in the mfsmaster.cfg file correct?

**Comment by chogata** [comment]
Yes, I meant mfsmaster.cfg of course, I should have mentioned that :)

**Reply by mandrew1983** [reply]
Hey thanks again for your help on this matter. It looks like the rebalancing has significantly sped up. I'm still monitoring it but it looks like everything is working out.

**Comment by mandrew1983** [comment]
I figured, I just wanted to confirm. I'll change the configuration and let
you know. Again, thank you very much for your assistance on this matter.

Andrew Ona
